### PR TITLE
8334085: Test crash: assert(thread->held_monitor_count() == 0) failed: Must be

### PR DESCRIPTION
**Pull request by sspitsyn**
The test `serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java` is failing with the assert in the `thaw_internal()` function. The assert is not fully correct as it does not account for an unexpected scenario.

Thanks to Patricio for reproducing this failure and identifying the root cause:
> The problem is that we can unmount a virtual thread, then mount it again, thaw a few frames, execute code that acquires a JNI monitor, and then call thaw again without releasing that monitor. In this test this will happen if the vthread is unmounted in System.out.println("Thread doing JNI call: " ...) because of contention with the main thread doing System.out.println("Main waiting for event.").
The issue can be reproduced by adding Thread.yield() before jniMonitorEnterAndLetObjectDie(). 

The fix corrects the assert to account for the `thread->jni_monitor_count()`.
Question: Is the same scenario possible for non-JNI monitors as well?
Also, the fix includes the test tweak described above which makes this failure always reproducible.

Testing:
 - Ran the test `GetOwnedMonitorInfoTest.java` locally
 - Mach5 tiers 1-6 are passed

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334085](https://bugs.openjdk.org/browse/JDK-8334085): Test crash: assert(thread-&gt;held_monitor_count() == 0) failed: Must be (**Bug** - P3)

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) 🔄 Re-review required (review applies to [9c10c555](https://git.openjdk.org/jdk/pull/20294/files/9c10c555ce6613b636ffadb6dd993b502f13e060))
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20294/head:pull/20294` \
`$ git checkout pull/20294`

Update a local copy of the PR: \
`$ git checkout pull/20294` \
`$ git pull https://git.openjdk.org/jdk.git pull/20294/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20294`

View PR using the GUI difftool: \
`$ git pr show -t 20294`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20294.diff">https://git.openjdk.org/jdk/pull/20294.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20294#issuecomment-2244406251)

### CODE REVIEW

**Review by pchilano — APPROVED** [review]
Looks good to me, thanks Serguei!

**Comment by sspitsyn** [comment]
Thank you for prompt review, Patricio!

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java [hunk 1]
```diff
@@ -86,6 +86,7 @@ public static void runTest(boolean isVirtual, boolean jni) throws Exception {
                 System.out.println("Thread doing JNI call: "
                                    + Thread.currentThread().getName());
 
+                Thread.yield();
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:89]: Please add a comment explaining why the yield is present

**Review by dholmes-ora — COMMENTED** [review]
Looks okay but one request.

Thanks

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java (hunk 1 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:91]: Thanks, David. Added a comment. Please, let me know if new comment is not good enough. I do not want to explain the whole scenario there.

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java [hunk 2]
```diff
@@ -86,6 +86,8 @@ public static void runTest(boolean isVirtual, boolean jni) throws Exception {
                 System.out.println("Thread doing JNI call: "
                                    + Thread.currentThread().getName());
 
+                // Extra unmount helps to reproduce 8334085.
+                // Two sub-sequential thaws are needed in that sceanrio.
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:90]: s/sceanario/scenario

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java (hunk 1 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:91]: The comment is sufficient - thanks

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java [hunk 3]
```diff
@@ -53,7 +53,7 @@ public class GetOwnedMonitorInfoTest {
     private static native boolean hasEventPosted();
 
     private static void jniMonitorEnterAndLetObjectDie() {
-        // The monitor iterator used by GetOwnedMonitorInfo used to
+        // The monitor iterator used by GetOwnedMonitorInfo to
```
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:56]: The original was: "used by GetOwnedMonitorInfo used to assert".
The word "used" is repeated twice.
Changed to: "used by GetOwnedMonitorInfo to assert".
Do I miss anything?

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java (hunk 2 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:90]: Sorry for the typo. Fixed now.

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java (hunk 1 above)
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java:91]: Thanks!

File: test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sspitsyn) [reply test/hotspot/jtreg/serviceability/jvmti/GetOwnedMonitorInfo/GetOwnedMonitorInfoTest.java]: Okay, thanks.  It is a little bit confusing. Restored the original comment.

**Review by dholmes-ora — APPROVED** [review]

**Comment by sspitsyn** [comment]
Thank you for review, David!

**Pull request closed without merging**